library/alloc/src/vec.rs

```diff
@@ -2508,17 +2508,19 @@ where
     }
 }
 
-impl<'a, T: 'a> SpecFromIter<&'a T, slice::Iter<'a, T>> for Vec<T>
-where
-    T: Copy,
-{
-    // reuses the extend specialization for T: Copy
+impl<'a, T: 'a + Copy> SpecFromIter<&'a T, slice::Iter<'a, T>> for Vec<T> {
+    #[cfg(not(test))]
     fn from_iter(iterator: slice::Iter<'a, T>) -> Self {
-        let mut vec = Vec::new();
-        // must delegate to spec_extend() since extend() itself delegates
-        // to spec_from for empty Vecs
-        vec.spec_extend(iterator);
-        vec
+        iterator.as_slice().to_vec()
```

Comment:
This can be called even on non-Copy T, so I'm not sure why we're specializing only for the Copy case here?

Reply:
I guess the theory is that spec_extend might have a more optimal code path? Regardless I would like a comment on this Copy impl indicating why it's not always used...

Reply:
I think the idea behind the pre-existing Vec specialization was to use the `TrustedLen` specialization for the T: Clone case and delegate to `extend` for `T: Copy` because extend in turn has a specialization that uses `ptr::copy_nonoverlapping`.

If `[T]:;to_vec()` now does all that work we can delegate there. The other way (always delegating `to_vec` to the `FromIterator`) probably would have worked too.

Reply:
I guess -- regardless -- it seems like these two cases are not all that dissimilar and presumably should share the same set of specializations rather than diverging; I would rather see this PR do that if possible.

Reply:
Sorry if I'm misunderstanding here, but `spec_extend`(and extend in general) and `to_vec` are sufficiently different in that extend has to reason about prior spare capacity(specifically `Vec::reserve()`) and use it if possible, whereas `to_vec` has more control of using exactly the right capacity, so it can all be allocated in one go at the start, so I chose the case which has more information about the current state because hopefully that would lead to better optimizations.

I can definitely expand the bound from `Copy` to `Clone`, not sure what perf impact that will have tho.

Reply:
Ah, interesting -- I for some reason thought the other specialization was for `collect()` not extend, in which case it'd also not have any prior capacity to deal with. (Though with the in-place collects I guess that's also somewhat untrue).

Let's try the bound change -- I think that's worth running through perf -- and if it doesn't cause us any problems (I would not expect it to, but not sure, haven't stared at this for too long), then we can merge a simpler version :)

Reply:
> Sorry if I'm misunderstanding here, but spec_extend(and extend in general) and to_vec are sufficiently different in that extend has to reason about prior spare capacity

When coming from a `FromIterator` implementation then it should know that length is 0 and eliminate most branches. `Vec::new().reserve(x)` and `Vec::with_capacity(x)` should have more or less the same performance.

>  I for some reason thought the other specialization was for collect() not extend, in which case it'd also not have any prior capacity to deal with

Uh, that usually is the case. `SpecFromIter` is for `collect()`. It just that some of the specializations sometimes happen to do `Vec::new().extend(iter)` (well `SpecExtend::spec_extend`) in some cases where collect and extend would do almost the same thing modulo a few branches the optimizer can eliminate. That way an extend specialization without duplicating it for the collect case.

Maybe now that the generic allocator support has landed I should move some of this stuff to a separate module and provide a better overview.

> I can definitely expand the bound from Copy to Clone (as that's the bound on to_vec), not sure what perf impact that will have tho.

That should be fine (assuming `min_specialization` won't complain) since `to_vec` now takes care of the Clone/Copy distinction.

If you do so please update the ASCII-art above the `SpecFrom` trait.

Reply:
I think relying on LLVM + optimizations to remove branches with 0 is fine usually, but here I think it's ok to be a bit more cautious to get rid of all possible allocations and whatnot. I think if all branches were being optimized out, it should have the same performance as this implementation, because at the end of the day they're doing the same thing.

I did find reading through all the specializations to be quite confusing, partially cause the names are all kinda similar in my head and when I see `Spec` I think `Spectral` not `Specialization`, so moving to a separate location with less naming conflicts would probably be good